models/alexnet/alexnet_impl.hpp

```diff
@@ -0,0 +1,126 @@
+/**
+ * @file alexnet_impl.hpp
+ * @author Shubham Agrawal
+ *
+ * Implementation of AlexNet using mlpack.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MODELS_MODELS_ALEXNET_ALEXNET_IMPL_HPP
+#define MODELS_MODELS_ALEXNET_ALEXNET_IMPL_HPP
+
+#include "alexnet.hpp"
+
+namespace mlpack {
+namespace models {
+
+template<typename MatType>
+AlexNetType<MatType>::AlexNetType() :
+    ann::MultiLayer<MatType>(),
+    numClasses(1000),
+    includeTop(true)
+{
+  MakeModel();
+}
+
+template<typename MatType>
+AlexNetType<MatType>::AlexNetType(
+    const size_t numClasses,
+    const bool includeTop) :
+    ann::MultiLayer<MatType>(),
+    numClasses(numClasses),
+    includeTop(includeTop)
+{
+  MakeModel();
+}
+
+template<typename MatType>
+AlexNetType<MatType>::AlexNetType(const AlexNetType& other) :
+    ann::MultiLayer<MatType>(other),
+    numClasses(other.numClasses),
+    includeTop(other.includeTop)
+{
+  // Nothing to do here.
+}
+
+template<typename MatType>
+AlexNetType<MatType>::AlexNetType(AlexNetType&& other) :
+    ann::MultiLayer<MatType>(std::move(other)),
+    numClasses(std::move(other.numClasses)),
+    includeTop(std::move(other.includeTop))
+{
+  // Nothing to do here.
+}
+
+template<typename MatType>
+AlexNetType<MatType>& AlexNetType<MatType>::operator=(const AlexNetType& other)
+{
+  if (this != &other)
+  {
+    ann::MultiLayer<MatType>::operator=(other);
+    numClasses = other.numClasses;
+    includeTop = other.includeTop;
+  }
+
+  return *this;
+}
+
+template<typename MatType>
+AlexNetType<MatType>& AlexNetType<MatType>::operator=(AlexNetType&& other)
+{
+  if (this != &other)
+  {
+    ann::MultiLayer<MatType>::operator=(std::move(other));
+    numClasses = std::move(other.numClasses);
+    includeTop = std::move(other.includeTop);
+  }
+
+  return *this;
+}
+
+template<typename MatType>
+template<typename Archive>
+void AlexNetType<MatType>::serialize(
+    Archive& ar, const uint32_t /* version */)
+{
+  ar(cereal::base_class<ann::MultiLayer<MatType>>(this));
+
+  ar(CEREAL_NVP(numClasses));
+  ar(CEREAL_NVP(includeTop));
+}
+
+template<typename MatType>
+void AlexNetType<MatType>::MakeModel()
+{
+  this->template Add<ann::Convolution>(64, 11, 11, 4, 4, 2, 2);
+  this->template Add<ann::ReLU>();
+  this->template Add<ann::MaxPooling>(3, 3, 2, 2);
+  this->template Add<ann::Convolution>(192, 5, 5, 1, 1, 2, 2);
+  this->template Add<ann::ReLU>();
+  this->template Add<ann::MaxPooling>(3, 3, 2, 2);
+  this->template Add<ann::Convolution>(384, 3, 3, 1, 1, 1, 1);
+  this->template Add<ann::ReLU>();
+  this->template Add<ann::Convolution>(256, 3, 3, 1, 1, 1, 1);
+  this->template Add<ann::ReLU>();
+  this->template Add<ann::Convolution>(256, 3, 3, 1, 1, 1, 1);
```

Comment:
Are these the right number of filters for each layer?  I am looking at https://proceedings.neurips.cc/paper/2012/file/c399862d3b9d6b76c8436e924a68c45b-Paper.pdf, Figure 2, and it seems off to me.  For instance, I believe the first layer should be a grouped convolution (with 2 groups), with 96 kernels (48 in each group), not 64.  Correct me if I overlooked something, or if I am looking in the wrong place.

Reply:
Idk why torchvision has a different architecture
https://github.com/pytorch/vision/blob/main/torchvision/models/alexnet.py
I don't think anybody is using Grouped Conv for AlexNet for some reason.

Reply:
Fair enough, if this matches the torchvision reference implementation that's good enough for me. :+1: I guess maybe that was done because I think grouped convolution was only used in AlexNet as a way to distribute across two GPUs.